modules/data.atmosphere/R/download.NOAA_GEFS.R

```diff
@@ -0,0 +1,260 @@
+    PEcAn.logger::logger.severe("Invalid dates: end date occurs before start date")
+  } else if (as.numeric(end_date - start_date, units="hours") < 6) { #Done separately to produce a more helpful error message.
+    PEcAn.logger::logger.severe("Times not far enough appart for a forecast to fall between them.  Forecasts occur every six hours; make sure start 
+                                and end dates are at least 6 hours appart.")
+  }
+  
+  #Set the end forecast date (default is the full 16 days)
+  if (end_date > start_date + lubridate::days(16)) {
+    end_date = start_date + lubridate::days(16)
+  }
+  
+  #Round the starting date/time down to the previous block of 6 hours.  Adjust the time frame to match.
+  forecast_hour = (lubridate::hour(start_date) %/% 6) * 6 #Integer division by 6 followed by re-multiplication acts like a "floor function" for multiples of 6
+  increments = as.integer(as.numeric(end_date - start_date, units = "hours") / 6) #Calculating the number of forecasts between start and end dates.
+  increments = increments + ((lubridate::hour(end_date) - lubridate::hour(start_date)) %/% 6) #These calculations are required to use the rnoaa package.
+  
+  end_hour = sprintf("%04d", ((forecast_hour + (increments * 6)) %% 24) * 100)  #Calculating the starting hour as a string, which is required type to access the 
+                                                                        #data via the rnoaa package
+  forecast_hour = sprintf("%04d", forecast_hour * 100)  #Having the end date as a string is useful later, too.
+  
+  #Recreate the adjusted start and end dates.
+  start_date = as.POSIXct(paste0(lubridate::year(start_date), "-", lubridate::month(start_date), "-", lubridate::day(start_date), " ", 
+                                 substring(forecast_hour, 1,2), ":00:00"), tz="UTC")
+  end_date = start_date + lubridate::hours(increments * 6)
+  
+  #Bounds date checking
+  #NOAA's GEFS database maintains a rolling 12 days of forecast data for access through this function.
+  #We do want Sys.Date() here - NOAA makes data unavaliable days at a time, not forecasts at a time.
+  NOAA_GEFS_Start_Date = as.POSIXct(Sys.Date(), tz="UTC") - lubridate::days(11)  #Subtracting 11 days is correct, not 12.
+  
+  #Check to see if start_date is valid. This must be done after date adjustment.
+  if (as.POSIXct(Sys.time(), tz="UTC") < start_date || start_date < NOAA_GEFS_Start_Date) {
+    PEcAn.logger::logger.severe(sprintf('Start date (%s) exceeds the NOAA GEFS range (%s to %s).',
+                                        start_date,
+                                        NOAA_GEFS_Start_Date, Sys.Date()))
+  }
+  
+  if (lubridate::hour(start_date) > 23) {
+    PEcAn.logger::logger.severe(sprintf("Start time %s is not a valid time", lubridate::hour(start_date)))
+  }
+  
+  if (lubridate::hour(end_date) > 23) {  #Done separately from the previous if statement in order to have more specific error messages.
```

Comment:
Shouldn't both of these be `>=`? Why can't I get data starting at 23:30 (11:30PM)?

Reply:
You can.  This statement only looks at the hour, not the minute.  So, as long as the hour is 23, you're good to go.  At 23:30, the hour is 23, so the error statement won't be triggered.